src/mbox/message-box.lisp

```diff
@@ -200,14 +236,26 @@ It will be apply'ed with the rest of the args when the message was 'popped' from
                      :time-out time-out
                      :handler-fun-args handler-fun-args
                      :handler-result 'no-result)))
-    (log:trace "~a: withreply: waiting for arrival of result..." (name msgbox))
-    (bt2:with-lock-held (withreply-lock)
-      (log:trace "~a: pushing item to queue: ~a" (name msgbox) push-item)
-      (queue:pushq queue push-item)
-
-      (if time-out
-          (wait-and-probe-for-msg-handler-result msgbox push-item)
-          (bt2:condition-wait withreply-cvar withreply-lock)))
+    (cond
+      (time-out
+       (bt2:with-lock-held (withreply-lock)
+         (log:trace "~a: pushing item to queue: ~a" (name msgbox) push-item)
+         (queue:pushq queue push-item)
+         (ensure-thread-is-running msgbox))
+
+       ;; It is important to leave lock withreply-lock
+       ;; before we will wait for result. Otherwisee handler-fun
+       ;; will not be able to do it's job:
+       (log:trace "~a: withreply: waiting for arrival of result..." (name msgbox))
+       (wait-and-probe-for-msg-handler-result msgbox push-item))
+      (t
+       (bt2:with-lock-held (withreply-lock)
+         (log:trace "~a: pushing item to queue: ~a" (name msgbox) push-item)
+         (queue:pushq queue push-item)
+         (ensure-thread-is-running msgbox)
```

Comment:
We also need to add this to `submit/no-reply`. And then, probably it's better to put it to `submit`, to catch both ways of submitting.